certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,240 @@
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushFront {
+    // Sanity
+    requireInvariant boundariesConsistency;
+    
+    bytes32 value;
+
+    int128 beginBefore = begin();
+    int128 beginNext = beginBefore - 1;
+    bytes32 frontBefore = front();
+
+    pushFront@withrevert(value);
+
+    int128 beginAfter = begin();
+    bytes32 frontAfter = front();
+
+    bool success = !lastReverted;
+
+    // liveness
+    assert success;
+
+    // effect
+    assert success => (
+        frontAfter == value && // Front set to value
+        beginAfter == beginNext // Begin is adjusted
+    ), "front set to value and begin is adjusted";
+
+    // no side effect
+    assert frontBefore == frontAfter <=> frontBefore == value, "front doesn't change only if it was already set";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: pushBack adds an element at the end of the queue                                              │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBack {
+    // Sanity
+    requireInvariant boundariesConsistency;
+
+    bytes32 value;
+
+    int128 endBefore = end();
+    int128 endNext = endBefore + 1;
+    bytes32 backBefore = back();
+
+    pushBack@withrevert(value);
+
+    int128 endAfter = end();
+    bytes32 backAfter = back();
+
+    bool success = !lastReverted;
+
+    // liveness
+    assert success;
+
+    // effect
+    assert success => (
+        backAfter == value && // Back set to true
+        endAfter == endNext // End is adjusted
+    ), "back set to value and begin is adjusted";
+
+    // no side effect
+    assert backBefore == backAfter <=> backBefore == value, "back doesn't change only if it was already set";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popFront removes an element from the beginning of the queue                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFront {
+    requireInvariant boundariesConsistency;
+    bytes32 value;
+    require !empty();
+
+    int128 beginBefore = begin();
+    int128 beginNext = beginBefore + 1;
+    bytes32 frontBefore = front();
+
+    popFront@withrevert();
+
+    int128 beginAfter = begin();
+    bytes32 frontAfter = front();
+
+    bool success = !lastReverted;
+
+    // liveness
+    assert success;
+
+    // effect
+    assert success => beginAfter == beginNext, "begin is adjusted ";
+
+    // no side effect
+    assert frontBefore == frontAfter <=> frontBefore == value, "front doesn't change only if it was already set";
```

Comment:
I don't get what you are checking here.

Reply:
This one is for checking that if the `front` didn't change, it can only be because the next value after popping the old `front` is equal to the older value